Source/HTTP/XCurl/XCurlEasyRequest.cpp

```diff
@@ -0,0 +1,342 @@
+        RETURN_IF_FAILED(HCHttpCallRequestGetHeaderAtIndex(hcCall, i, &name, &value));
+        RETURN_IF_FAILED(easyRequest->AddHeader(name, value));
+    }
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_HTTPHEADER, easyRequest->m_headers));
+
+    // timeout
+    uint32_t timeoutSeconds{ 0 };
+    RETURN_IF_FAILED(HCHttpCallRequestGetTimeout(hcCall, &timeoutSeconds));
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_TIMEOUT_MS, static_cast<long>(timeoutSeconds * 1000)));
+
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_VERBOSE, 0)); // verbose logging (0 off, 1 on)
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_HEADER, 0)); // do not write headers to the write callback
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_ERRORBUFFER, easyRequest->m_errorBuffer));
+
+    // write data callback
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_WRITEFUNCTION, &WriteDataCallback));
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_WRITEDATA, easyRequest.get()));
+
+    // write header callback
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_HEADERFUNCTION, &WriteHeaderCallback));
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_HEADERDATA, easyRequest.get()));
+
+    // debug callback
+    RETURN_IF_FAILED(easyRequest->SetOpt(CURLOPT_DEBUGFUNCTION, &DebugCallback));
+
+    return Result<HC_UNIQUE_PTR<XCurlEasyRequest>>{ std::move(easyRequest) };
+}
+
+CURL* XCurlEasyRequest::Handle() const noexcept
+{
+    return m_curlEasyHandle;
+}
+
+void XCurlEasyRequest::Complete(CURLcode result)
+{
+    HC_TRACE_VERBOSE(HTTPCLIENT, "XCurlEasyRequest::Complete: CURLCode=%ul", result);
+
+    long httpStatus = 0;
+    auto getInfoRes = curl_easy_getinfo(m_curlEasyHandle, CURLINFO_RESPONSE_CODE, &httpStatus);
+    if (getInfoRes != CURLE_OK)
+    {
+        result = getInfoRes;
+    }
+
+    if (result != CURLE_OK)
+    {
+        HC_TRACE_VERBOSE(HTTPCLIENT, "XCurlEasyRequest::m_errorBuffer='%s'", m_errorBuffer);
+
+        HRESULT hr = HCHttpCallResponseSetNetworkErrorCode(m_hcCallHandle, E_FAIL, result);
+        assert(SUCCEEDED(hr));
+
+        hr = HCHttpCallResponseSetPlatformNetworkErrorMessage(m_hcCallHandle, curl_easy_strerror(result));
+        assert(SUCCEEDED(hr));
+    }
+
+    HRESULT hr = HCHttpCallResponseSetStatusCode(m_hcCallHandle, httpStatus);
+    assert(SUCCEEDED(hr));
+    UNREFERENCED_PARAMETER(hr);
+
+    // Always complete with XAsync success; http/network errors returned via HCCallHandle
+    XAsyncComplete(m_asyncBlock, S_OK, 0);
+}
+
+void XCurlEasyRequest::Fail(HRESULT hr)
+{
+    HC_TRACE_ERROR_HR(HTTPCLIENT, hr, "XCurlEasyRequest::Fail");
+    XAsyncComplete(m_asyncBlock, hr, 0);
+}
+
+HRESULT XCurlEasyRequest::AddHeader(char const* name, char const* value) noexcept
+{
+    int required = std::snprintf(nullptr, 0, "%s: %s", name, value);
+    assert(required > 0);
+
+    m_headersBuffer.emplace_back();
+    auto& header = m_headersBuffer.back();
+
+    header.resize(static_cast<size_t>(required) + 1, '\0');
```

Comment:
the +1/-1 dance is not necessary: std::string guarantees that there is space for the null terminator and I've been taking advantage of that in newer code